src/rendering/webgl/programs/node.image.ts

```diff
@@ -104,41 +120,95 @@ export default function getNodeImageProgram(): typeof AbstractNodeImageProgram {
     const canvas = document.createElement("canvas");
     const ctx = canvas.getContext("2d") as CanvasRenderingContext2D;
 
-    canvas.width = pendingImages.reduce((iter, { size }) => iter + size, hasReceivedImages ? textureImage.width : 0);
-    canvas.height = Math.max(hasReceivedImages ? textureImage.height : 0, ...pendingImages.map(({ size }) => size));
+    // limit canvas size to avoid browser and platform limits
+    let totalWidth = hasReceivedImages ? textureImage.width : 0;
+    let totalHeight = hasReceivedImages ? textureImage.height : 0;
+
+    // initialize image drawing offsets with current write position
+    let xOffset = writePositionX;
+    let yOffset = writePositionY;
+
+    /**
+     * Draws a (full or partial) row of images into the atlas texture
+     * @param pendingImages
+     */
+    const drawRow = (pendingImages: PendingImage[]) => {
+      // update canvas size before drawing
+      if (canvas.width !== totalWidth || canvas.height !== totalHeight) {
+        canvas.width = Math.min(MAX_CANVAS_WIDTH, totalWidth);
+        canvas.height = totalHeight;
+
+        // draw previous texture into resized canvas
+        if (hasReceivedImages) {
+          ctx.putImageData(textureImage, 0, 0);
+        }
+      }
 
-    let xOffset = 0;
-    if (hasReceivedImages) {
-      ctx.putImageData(textureImage, 0, 0);
-      xOffset = textureImage.width;
-    }
-    pendingImages.forEach(({ id, image, size }) => {
-      const imageSizeInTexture = Math.min(MAX_TEXTURE_SIZE, size);
-
-      // Crop image, to only keep the biggest square, centered:
-      let dx = 0,
-        dy = 0;
-      if ((image.width || 0) > (image.height || 0)) {
-        dx = (image.width - image.height) / 2;
-      } else {
-        dy = (image.height - image.width) / 2;
+      pendingImages.forEach(({ id, image, size }) => {
+        const imageSizeInTexture = Math.min(MAX_TEXTURE_SIZE, size);
+
+        // Crop image, to only keep the biggest square, centered:
+        let dx = 0,
+          dy = 0;
+        if ((image.width || 0) > (image.height || 0)) {
```

Comment:
This `|| 0` seems to imply `width` and `height` could not be set, in which case the `else` branch could produce a `NaN`, no?

Reply:
Fair enough this code is not yours but the diff was misleading.